director/jobs/models.py

```diff
@@ -0,0 +1,433 @@
+        related_name="jobs",
+        help_text="The project this job is associated with.",
+    )
+
+    creator = models.ForeignKey(
+        User,
+        null=True,  # Should only be null if the creator is deleted
+        blank=True,
+        on_delete=models.SET_NULL,
+        related_name="jobs_created",
+        help_text="The user who created the job.",
+    )
+
+    created = models.DateTimeField(
+        null=True, auto_now_add=True, help_text="The time the job was created."
+    )
+
+    zone = models.ForeignKey(
+        Zone,
+        null=True,
+        blank=True,
+        on_delete=models.SET_NULL,
+        related_name="jobs",
+        help_text="The project this job is associated with.",
+    )
+
+    queue = models.CharField(
+        max_length=64,
+        blank=True,
+        null=True,
+        help_text="The identifier of the queue the job was posted to.",
+    )
+
+    began = models.DateTimeField(null=True, help_text="The time the job began.")
+    ended = models.DateTimeField(null=True, help_text="The time the job ended.")
+    status = models.CharField(
+        max_length=32,
+        choices=JobStatus.as_choices(),
+        blank=True,
+        null=True,
+        help_text="The current status of the job.",
+    )
+
+    method = models.CharField(
+        max_length=32, choices=JobMethod.as_choices(), help_text="The job method."
+    )
+    params = FallbackJSONField(
+        blank=True, null=True, help_text="The parameters of the job; a JSON object."
+    )
+    result = FallbackJSONField(
+        blank=True, null=True, help_text="The result of the job; a JSON value."
+    )
+    log = FallbackJSONField(
+        blank=True,
+        null=True,
+        help_text="The job log; a JSON array of log objects, including any errors.",
+    )
+    runtime = models.FloatField(
+        blank=True, null=True, help_text="The running time of the job."
+    )
+
+    # Not a URLField because potential for non-standard schemes e.g. ws://
+    url = models.CharField(
+        max_length=256,
+        blank=True,
+        null=True,
+        help_text="The URL of the job on the local network; can be used to interact with it.",
+    )
+
+    users = models.ManyToManyField(
+        User,
+        help_text="The users who have connected to the job; not necessarily currently connected.",
+    )
+
+    worker = models.CharField(
+        max_length=64,
+        blank=True,
+        null=True,
+        help_text="The identifier of the worker that ran the job.",
+    )
+    retries = models.IntegerField(
+        blank=True, null=True, help_text="The number of retries to fulfil the job.",
+    )
+
+    @property
+    def position(self):
+        """
+        Position of the job in the queue.
+
+        Counts the number of jobs that were `created` before this
+        job and that have no `ended` date.
+        """
+        return (
+            Job.objects.filter(created__lt=self.created, ended__isnull=True).count() + 1
```

Comment:
Since this is a "relative" count the `+1` is redundant. I know this is the stupidest nit-picky comment ever.